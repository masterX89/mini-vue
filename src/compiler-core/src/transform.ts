import { NodeTypes } from './ast'
import { helperMapName, TO_DISPLAY_STRING } from './runtimeHelpers'

export function transform(root, options = {}) {
  const context = createTransformerContext(root, options)
  traverseNode(root, context)
  createRootCodegen(root)
  root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root: any) {
  const { children } = root
  if (children.length === 1) {
    const child = root.children[0]
    if (child.type === NodeTypes.ELEMENT) {
      const codegenNode = child.codegenNode
      root.codegenNode = codegenNode
    } else {
      root.codegenNode = child
    }
  }
}

function traverseNode(node: any, context) {
  const nodeTransforms = context.nodeTransforms
  const exitFns: any = []
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i]
    const onExit = transform(node, context)
    if (onExit) {
      exitFns.push(onExit)
    }
  }
  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING)
      break
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context)
      break
    default:
      break
  }
  let i = exitFns.length
  while (i--) {
    exitFns[i]()
  }
}

function traverseChildren(node, context) {
  const children = node.children
  for (let i = 0; i < children.length; i++) {
    traverseNode(children[i], context)
  }
}

function createTransformerContext(root: any, options: any = {}) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1)
    },
  }
  return context
}
