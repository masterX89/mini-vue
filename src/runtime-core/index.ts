export { h } from './h'
export { renderSlot } from './helpers/renderSlot'
export { createTextVNode } from './vnode'
export { getCurrentInstance } from './component'
export { provide, inject } from './apiInject'
export { createRenderer } from './renderer'
export { nextTick } from './scheduler'
export * from '../reactivity'
