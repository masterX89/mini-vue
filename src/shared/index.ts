export const extend = Object.assign

export const isObject = (val) => val !== null && typeof val === 'object'
export const isString = (val) => typeof val === 'string'
export const isArray = (val) => Array.isArray(val)

// NaN 算作变更
export const hasChanged = (newVal, oldVal) => !Object.is(newVal, oldVal)
