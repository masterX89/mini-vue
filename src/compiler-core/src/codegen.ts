import { isString } from '../../shared'
import { NodeTypes } from './ast'
import {
  CREATE_ELEMENT_VNODE,
  helperMapName,
  TO_DISPLAY_STRING,
} from './runtimeHelpers'

export function generate(ast) {
  const context = generateContext()
  const { push } = context

  genFunctionPreamble(ast, context)
  const functionName = 'render'
  const args = ['_ctx', '_cache']
  const signature = args.join(', ')

  push(`function ${functionName}(${signature}) {`)
  push('return ')
  genNode(ast.codegenNode, context)
  push('}')
  return {
    code: context.code,
  }
}

function genFunctionPreamble(ast, context) {
  const { push } = context
  const VueBinging = 'Vue'
  const aliasHelper = (s) => `${helperMapName[s]}: _${helperMapName[s]}`
  if (ast.helpers.length > 0) {
    push(`const { ${ast.helpers.map(aliasHelper).join(', ')} } = ${VueBinging}`)
  }
  push('\n')
  push('return ')
}

function genNode(node, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context)
      break
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context)
      break
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break
    case NodeTypes.ELEMENT:
      genElement(node, context)
      break
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context)
      break
    default:
      break
  }
}

function genCompoundExpression(node: any, context: any) {
  const { children } = node
  const { push } = context
  for (let i = 0; i < children.length; i++) {
    const child = children[i]
    if (isString(child)) {
      push(child)
    } else {
      genNode(child, context)
    }
  }
}

function genElement(node, context) {
  const { push, helper } = context
  const { tag, props, children } = node
  push(`${helper(CREATE_ELEMENT_VNODE)}(`)
  genNodeList(genNullableArgs([tag, props, children]), context)
  push(')')
}

function genChildren(nodes, context) {
  if (nodes.length === 1) {
    const child = nodes[0]
    const type = child.type
    if (type === NodeTypes.TEXT || type === NodeTypes.INTERPOLATION || type === NodeTypes.COMPOUND_EXPRESSION) {
      genNodeList(nodes, context)
    } else {
      genNodeListAsArray(nodes, context)
    }
  } else {
    genNodeListAsArray(nodes, context)
  }
}

function genNodeListAsArray(nodes, context) {
  const { push } = context
  push('[')
  genNodeList(nodes, context)
  push(']')
}

function genNodeList(nodes, context) {
  const { push } = context
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i]
    if (isString(node)) {
      // tag or null props
      // FIXME: 可能来自 node 也可能来自 node.codegenNode 需要做统一
      if (node.slice(0, 1) === '"' || node === 'null') {
        push(node)
      } else {
        push(`"${node}"`)
      }
    } else if (Array.isArray(node)) {
      // Array: children nodes
      genChildren(node, context)
    } else {
      // Object: node
      genNode(node, context)
    }
    if (i < nodes.length - 1) {
      push(', ')
    }
  }
}

function genNullableArgs(args) {
  return args.map((arg) => arg || 'null')
}

function genExpression(node: any, context: any) {
  const { push } = context
  push(`${node.content}`)
}

function genInterpolation(node: any, context: any) {
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(')')
}

function genText(node, context) {
  const { push } = context
  push(`"${node.content}"`)
}

function generateContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    },
    helper(key) {
      return `_${helperMapName[key]}`
    },
  }
  return context
}
