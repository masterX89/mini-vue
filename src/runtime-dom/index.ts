import { createRenderer } from '../runtime-core'
import { isOn } from '../shared'

function createElement(type) {
  return document.createElement(type)
}

function patchProp(el, key, val) {
  if (isOn(key)) {
    const event = key.substring(2).toLowerCase()
    el.addEventListener(event, val)
  } else if (val === undefined || val === null) {
    el.removeAttribute(key)
  } else {
    el.setAttribute(key, val)
  }
}

function insert(el, container) {
  container.append(el)
}

const renderer: any = createRenderer({
  createElement,
  patchProp,
  insert,
})

export function createApp(...args) {
  return renderer.createApp(...args)
}

export * from '../runtime-core'
