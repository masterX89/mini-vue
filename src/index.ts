import { baseCompile } from './compiler-core/src'
import * as runtimeDom from './runtime-dom'
import { registerRuntimeCompiler } from './runtime-core'

export * from './runtime-dom'

function compileToFunction(template) {
  const { code } = baseCompile(template)
  // tmpl : const { toDisplayString: _toDisplayString} = Vue -> runtimeDom
  const render = new Function('Vue', code)(runtimeDom)
  return render
}

registerRuntimeCompiler(compileToFunction)
