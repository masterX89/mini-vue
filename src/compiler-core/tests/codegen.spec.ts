import { baseCompile } from '../src/compile'

describe('code generate', () => {
  it('should generate string', () => {
    const { code } = baseCompile('hi')
    expect(code).toMatchSnapshot()
  })

  it('should generate interpolation', () => {
    const { code } = baseCompile('{{message}}')
    expect(code).toMatchSnapshot()
  })

  it('should generate nested element', () => {
    const { code } = baseCompile('<div>hi, {{message}}</div>')
    expect(code).toMatchSnapshot()
  })

  it('should generate paragraph nested element', () => {
    const { code } = baseCompile('<div><p>hello</p><p>{{msg}}</p></div>')
    expect(code).toMatchSnapshot()
  })
})
