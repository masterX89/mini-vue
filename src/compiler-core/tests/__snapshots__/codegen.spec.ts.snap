// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code generate should generate interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`code generate should generate nested element 1`] = `
"const { toDisplayString: _toDisplayString, createElementVNode: _createElementVNode } = Vue
return function render(_ctx, _cache) {return _createElementVNode(\\"div\\", null, \\"hi, \\" + _toDisplayString(_ctx.message))}"
`;

exports[`code generate should generate paragraph nested element 1`] = `
"const { createElementVNode: _createElementVNode, toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache) {return _createElementVNode(\\"div\\", null, _createElementVNode(p, null, \\"hello\\"), \\",\\", _createElementVNode(p, null, _toDisplayString(_ctx.msg)))}"
`;

exports[`code generate should generate string 1`] = `
"
return function render(_ctx, _cache) {return \\"hi\\"}"
`;
